app/actions/v3/service_instance_update_managed.rb

```diff
@@ -23,10 +23,16 @@ def initialize(instance, message, user_audit_info, audit_hash)
         @audit_hash = audit_hash
       end
 
-      def precursor
-        raise_if_cannot_proceed!
+      def preflight!
+        raise_if_invalid_update!
+        raise_if_renaming_shared_service_instance!
+        raise_if_invalid_plan_change!
+        raise_if_invalid_maintenance_info_change!
+        raise_if_cannot_update!
+      end
 
-        unless is_deleting?(service_instance) || !only_metadata?
+      def try_update_sync
+        if update_metadata_only?
```

Comment:
This is definitely improved. There are some questions that occur to me that are hard to understand, but perhaps could be made clearer with better names:
- if a service instance is being deleted, why is it that I can't change the metadata, but I can change the metadata and the name together as long as `allow_context_updates` is false?
- why is the audit event reported for the original service instance and not the updated one? But when only metadata is changed, it's reported on the updated service instance?

I'm sure there are good reasons, it's just that for someone unfamiliar with the code, it's not obvious. But perhaps the thing to do is merge this and improve it later.

Reply:
Both questions that I have had to think about in the code.

> If a service instance is being deleted, why is it that I can't change the metadata, but I can change the metadata and the name together as long as allow_context_updates is false?

I think there is some more subtlety here that was not captured either in the previous implementation and neither here. When changing metadata we made the decision to allow it only while create and update is in progress for all resources. In that sense it's consistent. I guess the idea is that while delete is happening it might be more confusing to accept any other operation. 
Any other operation makes the state check first and will fail if another operation is in progress - delete in this case. So name and metadata update that could happen later doesn't really happen. 
```
 lock = UpdaterLock.new(service_instance)
 lock.lock!
```
will raise when other operations are in progress.

> why is the audit event reported for the original service instance and not the updated one? But when only metadata is changed, it's reported on the updated service instance?

My understanding is that the event log does not include metadata so it is not relevant which instance is used. However, I will change it to be the updated one just for consistency.